client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,314 @@
+			.ok_or("Extrinsics not found".to_string())?;
+		tracing::debug!(target: "state_tracing", "Found {} extrinsics", extrinsics.len());
+		let parent_hash = header.parent_hash().clone();
+		let parent_id = BlockId::Hash(parent_hash.clone());
+		// Pop digest else RuntimePanic due to: 'Number of digest items must match that calculated.'
+		header.digest_mut().pop();
+		let block = Block::new(header, extrinsics);
+
+		let targets = if let Some(t) = &self.targets { t } else { DEFAULT_TARGETS };
+		let storage_keys = if let Some(s) = &self.storage_keys {
+			s
+		} else {
+			DEFAULT_STORAGE_KEYS
+		};
+		let spans = Mutex::new(HashMap::new());
+		let events = Mutex::new(Vec::new());
+		let block_subscriber = BlockSubscriber::new(targets, spans, events);
+		let dispatch = Dispatch::new(block_subscriber);
+
+		{
+			let dispatcher_span = tracing::debug_span!(target: "state_tracing",
+				"execute_block", extrinsics_len = block.extrinsics().len());
+			let _guard = dispatcher_span.enter();
+			if let Err(e) = dispatcher::with_default(&dispatch, || {
+				let span = tracing::info_span!(
+					target: TRACE_TARGET,
+					"trace_block",
+				);
+				let _enter = span.enter();
+				self.client.runtime_api().execute_block(&parent_id, block)
+			}) {
+				return Err(format!("Error executing block: {:?}", e));
+			}
+		}
+
+		let block_subscriber = dispatch.downcast_ref::<BlockSubscriber>()
+			.ok_or("Cannot downcast Dispatch to BlockSubscriber after tracing block")?;
+		let spans: Vec<_> = block_subscriber.spans
+			.lock()
+			.drain()
+			.map(|(_, s)| SpanDatum::from(s))
+			// Patch wasm identifiers
+			.filter_map(|s| patch_and_filter(s, targets))
+			.collect();
+		let events: Vec<_> = block_subscriber.events
+			.lock()
+			.drain(..)
+			.filter(|e| event_key_filter(e, storage_keys))
+			.map(|s| s.into())
+			.collect();
+		tracing::debug!(target: "state_tracing", "Captured {} spans and {} events", spans.len(), events.len());
+
+		let block_trace = BlockTrace {
+			block_hash: block_id_as_string(id),
+			parent_hash: block_id_as_string(parent_id),
+			tracing_targets: targets.to_string(),
+			storage_keys: storage_keys.to_string(),
+			spans,
+			events,
+		};
+		let block_trace_size = serde_json::to_vec(&block_trace)
+			.map_err(|e| format!("Failed to serialize payload: {}", e))?
+			.len();
+		let response = if block_trace_size  > MAX_PAYLOAD - BASE_RPC_PAYLOAD {
+			TraceBlockResponse::TraceError(TraceError {
+				error:
+					format!("Payload was {} bytes; it must be less than {} bytes",
+						block_trace_size, MAX_PAYLOAD - BASE_RPC_PAYLOAD,
+					)
+			})
+		} else {
+			TraceBlockResponse::BlockTrace(block_trace)
+		};
+
+		Ok(response)
+	}
+}
+
+fn event_key_filter(event: &TraceEvent, storage_keys: &str) -> bool {
+	event.values.string_values.get("key")
+		.and_then(|key| Some(check_target(storage_keys, key, &event.level)))
+		.unwrap_or(false)
+}
+
+fn patch_and_filter(mut span: SpanDatum, targets: &str) -> Option<Span> {
```

Comment:
Can we document this function? It's not obvious to me what this does and what it's side-effects has for impact.

Such as: 
if the `WASM_NAME_KEY` is removed the span.name is set to that
if the `WASM_TARGET_NAME` is removed the span.target is set to that

Reply:
@mattrutherford can you put a comment on this to explain some background on the the special wasm targets.

Reply:
btw I put in a comment above this